### PR TITLE
set find_unused_parameters=False in DDP as in pytorch

### PR DESCRIPTION
**Pull request by awaelchli**
Lightning currently forces `find_unused_parameters=True` in DDP, but pytorch recommends False.
Context: brief discussion here: #5185, suggested by @ananthsub

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -58,6 +58,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed `iou` [func] to allow float input ([#4704](https://github.com/PyTorchLightning/pytorch-lightning/pull/4704))
 
 
+- `stat_scores` metric now calculates stat scores over all classes and gains new parameters, in line with the new `StatScores` metric ([#4839](https://github.com/PyTorchLightning/pytorch-lightning/pull/4839))
+
+
```
Comment (ananthsub) [inline CHANGELOG.md:63]: rebase issue?

**Review by ananthsub — APPROVED** [review]
thanks for splitting this out @awaelchli !

File: CHANGELOG.md (hunk 1 above)
Comment (awaelchli) [reply CHANGELOG.md]: yes, when cherry picking some blueberries made it into the basket :)

**Comment by SeanNaren** [comment]
With this change we might see some models breaking that had parameters which were unused being pruned out automatically for them. I think @tchaton ran into a test breaking due to this as well!

@blefaudeux spotted iGPT was working due to find_unused_parameters being set to True by default I think: https://github.com/teddykoker/image-gpt

I think this is fine, as long as we make it clear that this change is being made + having an easy way to turn it back on. Currently the user would need to instantiate the ddp plugin to enable this flag, anyway we could potentially add this to the trainer?

File: tests/models/test_sync_batchnorm.py [hunk 2]
```diff
@@ -107,6 +108,7 @@ def test_sync_batchnorm_ddp(tmpdir):
         sync_batchnorm=True,
         num_sanity_val_steps=0,
         replace_sampler_ddp=False,
+        plugins=[DDPPlugin(find_unused_parameters=True)]
```
Comment (awaelchli) [inline tests/models/test_sync_batchnorm.py:111]: had to set this here to true, as mentioned by @SeanNaren in the comment above. 
turns out this modification to the test is already in master branch

**Comment by awaelchli** [comment]
@SeanNaren sounds like the reason we had it set to True before was simply so that users don't run into error messages despite it having a performance hit. 

I wonder how many users would have breaking code because of this. Perhaps we should print deprecation warnings for this change, the question is just how to do show it only to the users who would need it and not just everybody.

**Review by carmocca — APPROVED** [review]

**Comment by SeanNaren** [comment]
> @SeanNaren sounds like the reason we had it set to True before was simply so that users don't run into error messages despite it having a performance hit.
> 
> I wonder how many users would have breaking code because of this. Perhaps we should print deprecation warnings for this change, the question is just how to do show it only to the users who would need it and not just everybody.

This is going to be tricky, I think if it breaks it's better to hard crash so the user know what's happening. Then exposing the parameter `find_unused_parameters` to the trainer for users to enable. The overhead from using it from what I know is quite high, so we should turn this off by default for the users that do not need to worry about this.

File: tests/models/test_sync_batchnorm.py (hunk 2 above)
Comment (SeanNaren) [reply tests/models/test_sync_batchnorm.py:111]: Yeah for a test this is fine, but for the user you can't enable this unless you pass your own plugin, which is why I don't like this

Comment (awaelchli) [reply tests/models/test_sync_batchnorm.py:111]: yes, it's also not agnostic to command line input / device without directly. A new trainer flag is probably best for user.

**Comment by awaelchli** [comment]
sounds good. then do we want to do it directly in this PR here?

**Comment by carmocca** [comment]
> I wonder how many users would have breaking code because of this

I don't think it is a large percentage. And in any case, wouldn't most of them want to fix their unused parameters?

I don't like the idea of introducing  another Trainer parameter for this. It is just noise for everybody else. I would focus on making clear in the docs what is happening if the program fails due to the new behaviour and how to fix it. After all, it's not much harder to do:

`Trainer(plugins=[DDPPlugin(find_unused_parameters=True)]` than `Trainer(find_ddp_unused_parameters=True)`

**Comment by SeanNaren** [comment]
> > I wonder how many users would have breaking code because of this
> 
> I don't think it is a large percentage. And in any case, wouldn't most of them want to fix their unused parameters?
> 
> I don't like the idea of introducing another Trainer parameter for this. It is just noise for everybody else. I would focus on making clear in the docs what is happening if the program fails due to the new behaviour and how to fix it. After all, it's not much harder to do:
> 
> `Trainer(plugins=[DDPPlugin(find_unused_parameters=True)]` than `Trainer(find_ddp_unused_parameters=True)`

This requires' making the change within the code though, that's what I'm concerned about. If we feel confident that this won't break enough user code to add to the trainer args for convenience, then I'm happy :) more interested in getting this fix in ASAP.

Regardless let's do it in a separate PR, we can chat offline!

**Review by SeanNaren — APPROVED** [review]

**Comment by awaelchli** [comment]
alright, will open this for review then

**Review by tchaton — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@awaelchli mind resolve conflicts...

**Pull request merged**